<launch>

  <!-- args corresponding to environment variables -->
  <arg name="UAV_NAME" default="$(env UAV_NAME)" />
  <arg name="RUN_TYPE" default="$(env RUN_TYPE)" />
  <arg name="BROADCAST_IP" default="$(optenv BROADCAST_IP 192.168.0.255)" />

    <!-- other args -->
  <arg name="service_response_timeout" default="3" />
  <arg name="service_call_timeout" default="0.2" />
  <arg name="service_call_repeats" default="3" />
  <arg name="service_server_port" default="6000" />
  <arg name="topic_server_port" default="17001" />
  <arg name="broadcast_addr" default="$(arg BROADCAST_IP)" />

  <arg name="custom_config_uav_names" default="" />
  <arg name="custom_config" default="" />

  <group ns="$(arg UAV_NAME)">

    <!-- Service client -->
    <node name="service_udp_client" pkg="nimbro_service_transport" type="udp_client" output="screen">
      <param name="destination_port" value="$(arg service_server_port)" />
      <param name="response_timeout" value="$(arg service_response_timeout)" />
      <param name="call_timeout" value="$(arg service_call_timeout)" />
      <param name="call_repeats" value="$(arg service_call_repeats)" />
      <param name="topic_prefix" value="$(arg UAV_NAME)" />

        <!-- Load a config YAML file -->
      <rosparam if="$(eval arg('RUN_TYPE') == 'uav')" file="$(find mrs_general)/config/uav_names.yaml" />
      <rosparam if="$(eval arg('RUN_TYPE') == 'simulation')" file="$(find mrs_general)/config/uav_names_simulation.yaml" />
      <rosparam if="$(eval not arg('custom_config_uav_names') == '')" file="$(arg custom_config_uav_names)" />
      <rosparam command="load" file="$(find mrs_general)/config/communication_config.yaml" />
      <rosparam command="load" if="$(eval arg('custom_config') != '')" file="$(arg custom_config)" />
    </node>

    <!-- Service server -->
    <node name="service_udp_server" pkg="nimbro_service_transport" type="udp_server" output="screen">
      <param name="port" value="$(arg service_server_port)" />
      <param name="call_timeout" value="$(arg service_call_timeout)" />
      <param name="call_repeats" value="$(arg service_call_repeats)" />
    </node>

    <!-- Topic receiver -->
    <node name="topic_udp_receiver" pkg="nimbro_topic_transport" type="udp_receiver" output="screen">
      <!-- The port to receive packets on -->
      <param name="port" value="$(arg topic_server_port)" />
      <param name="topic_prefix" value="$(arg UAV_NAME)" />
    </node>

    <!-- Topic sender -->
    <node name="topic_udp_sender" pkg="nimbro_topic_transport" type="udp_sender" output="screen">

      <!-- The destination host name or IP address -->
      <param name="destination_addr" value="$(arg broadcast_addr)" />
      <param name="destination_port" value="$(arg topic_server_port)" />
      <param name="topic_prefix" value="$(arg UAV_NAME)" />

        <!-- Load a config YAML file -->
      <rosparam command="load" file="$(find mrs_general)/config/communication_config.yaml" />
    </node>
  </group>

</launch>
